kernel/src/hil/display.rs

```diff
@@ -0,0 +1,376 @@
+    /// - `INVAL`: Write is invalid or length is wrong.
+    /// - `BUSY`: Another write is in progress.
+    fn write(
+        &self,
+        buffer: &'static mut [u8],
+        len: usize,
+        reset_position: bool,
+    ) -> Result<(), ErrorCode>;
+
+    /// Flush the framebuffer changes to the hardware device.
+    ///
+    /// Some framebuffers keep in a temporary memory the changes and require a flush command
+    /// to send the changes to the hardware.
+    ///
+    /// Return values:
+    /// - `Ok(())`: Flush is in progress and the client will receive
+    ///    a call to `command_complete`.
+    /// - `ENOSUPPORT`: Flush has been done synchronous or there is no
+    ///    no need to flush the framebuffer.
+    /// - `BUSY`: Another write or flush is in progress.
+    fn flush(&self) -> Result<(), ErrorCode>;
+
+    /// Set the object to receive the asynchronous command callbacks.
+    fn set_client(&self, client: Option<&'a dyn FrameBufferClient>);
+}
+
+pub trait FrameBufferClient {
+    /// The framebuffer will call this function to notify that the write command has finished.
+    /// This is different from `command_complete` as it has to pass back the write buffer
+    fn write_complete(&self, buffer: &'static mut [u8], r: Result<(), ErrorCode>);
+
+    /// The framebuffer will call this function to notify that a command (except `write` and
+    /// `write_continue`) has finished.
+    fn command_complete(&self, r: Result<(), ErrorCode>);
+}
+
+pub trait FrameBufferSetup<'a>: FrameBuffer<'a> {
+    /// Sets the screen resolution (in pixels). Returns ENOSUPPORT if the resolution is
+    /// not supported. The function should return Ok(()) if the request is registered
+    /// and will be sent to the screen.
+    /// Upon Ok(()), the caller has to wait for the `command_complete` callback function
+    /// that will return the actual Result<(), ErrorCode> after setting the resolution.
+    fn set_mode(&self, mode: GraphicsMode) -> Result<(), ErrorCode>;
+
+    /// Returns the number of the resolutions supported.
+    /// should return at least one (the current resolution)
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen (most screens do not support such a request,
+    /// resolutions are described in the data sheet).
+    ///
+    /// If the screen supports such a feature, the driver should request this information
+    /// from the screen upfront.
+    fn get_num_supported_modes(&self) -> usize;
+
+    /// Can be called with an index from 0 .. count-1 and will
+    /// a tuple (width, height) with the current resolution (in pixels).
+    /// note that width and height may change due to rotation
+    ///
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    fn get_supported_mode(&self, index: usize) -> Option<GraphicsMode>;
```

Comment:
Why not return a slice?

Reply:
This was modeled initially to be sent over the syscall ABI which did no support a slice. You might be right, and the driver may be able to have an array with these modes and return a slice to it, I'll think about it.